### PR TITLE
WIP: Support creating XPC services

### PR DESCRIPTION
**Pull request by steven-joruk**
**The round trip tests have some intermittent failures**, so it's still WIP.

It's been a while since I started this PR. I need to fix up a couple of the tests that rely on the echo server but it's ready to start reviwing.

I've renamed XpcConnection to XpcClient for a better distinction from XpcListener, and implemented Stream for both of them so they're self-contained. This is most beneficial when returning a complete setup client or listener which was sent inside a Message.

## Future work

- Rework the API to prevent sending invalid messages, i.e. don't allow `send_message` to send a `Message::String` outside of a `Message::Dictionary`, it isn't allowed and your program will be killed.
- Determine whether a listener `xpc_connection_t` can be sent across a connection, then either drop the `Message::Client` / `Message::Listener` separation, or implement `Message::Listener`.
- Example unprivileged helper tool
- Look through the Apple docs for missing functionality (send with reply, barrier, transaction, etc.)
- Example client to client communication by sending a connection via a listener
- Example privileged helper tool, installed with SMJobBless

### CODE REVIEW

File: examples/echo-server/com.example.echo.plist [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+	<string>com.example.echo</string>
+	<key>MachServices</key>
+	<dict>
+		<key>com.example.echo</key>
+		<true/>
+	</dict>
+	<key>Program</key>
+        <string>/Library/PrivilegedHelperTools/com.example.echo</string>
+        <key>StandardOutPath</key>
+        <string>/var/log/WIP.log</string>
```
Comment (steven-joruk) [inline examples/echo-server/com.example.echo.plist]: Change this file name

File: xpc-connection/tests/message_round_trip.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
+use futures::{executor::block_on, StreamExt};
+use std::{collections::HashMap, error::Error, ffi::CString};
+use xpc_connection::{Message, XpcClient};
+
+/// FIXME: Outgoing strings, both as dictionary keys and as string values, are
```
Comment (steven-joruk) [inline xpc-connection/tests/message_round_trip.rs]: Remove this FIXME, it's fixed

File: examples/echo-server/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+authors = ["Steven Joruk <steven@joruk.com>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+embed_plist = "1"
+futures = { version = "0.3" }
+tokio = { version = "1", features = ["macros", "rt-multi-thread"] }
+xpc-connection = { path = "../../xpc-connection" }
```
Comment (steven-joruk) [inline examples/echo-server/Cargo.toml]: new line

File: examples/client/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+version = "0.1.0"
+authors = ["Steven Joruk <steven@joruk.com>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+futures = { version = "0.3" }
+tokio = { version = "1", features = ["macros", "rt-multi-thread"] }
+xpc-connection = { path = "../../xpc-connection" }
```
Comment (steven-joruk) [inline examples/client/Cargo.toml]: new line

File: examples/client/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "client"
+version = "0.1.0"
+authors = ["Steven Joruk <steven@joruk.com>"]
+edition = "2018"
+
```
Comment (dfrankland) [inline examples/client/Cargo.toml:7]: You may want to add the `publish = false` field here

https://doc.rust-lang.org/cargo/reference/manifest.html#the-publish-field

File: examples/echo-server/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "echo-server"
+version = "0.1.0"
+authors = ["Steven Joruk <steven@joruk.com>"]
+edition = "2018"
+
```
Comment (dfrankland) [inline examples/echo-server/Cargo.toml:7]: Same here

File: xpc-connection/Cargo.toml [hunk 7]
```diff
@@ -14,3 +14,6 @@ categories = ["os", "api-bindings", "concurrency", "encoding"]
 block = "0.1.6"
 futures = "0.3.4"
 xpc-connection-sys = { path="../xpc-connection-sys", version="0.1.0" }
+
+[dev-dependencies]
+tokio = { version = "1", features = ["macros", "rt-multi-thread"] }
```
Comment (dfrankland) [inline xpc-connection/Cargo.toml]: Missing line ending

File: examples/client/Cargo.toml (hunk 5 above)
Comment (steven-joruk) [reply examples/client/Cargo.toml:7]: Oh yeah, thanks

**Review by dfrankland — APPROVED** [review]
Nice! You've done a really amazing job. We can merge whenever you'd like

**Comment by steven-joruk** [comment]
Thanks :). It's not quite ready yet - some of the tests that use the echo server are intermittently failing. It's been a while since I did anything on this branch, I'll try to figure it out this week.

**Comment by steven-joruk** [comment]
I had a quick look last night, it looks like the cause is a combination of:
* Connections received by the `XpcListener`'s event handler aren't explicitly retained.
* There are different requirements for retaining a client vs a listener's client
* `Connection::drop` calls cancel then release, but cancel is async. I think I need to:
  * set up a new handler
  * cancel
  * release in the new handler block

**Comment by steven-joruk** [comment]
This last push fixes a memory leak due to creating a new dispatch queue and not releasing it, I've just made it use the default queue now.

It also fixes the issue with the echo server tests, all connections should now be freed, even if they're in the middle of their event handler when the `Drop` occurs.

**Review by dfrankland — APPROVED** [review]

**Pull request merged**